test/Microsoft.Azure.Devices.Edge.Test.Common/LeafDevice.cs

```diff
@@ -0,0 +1,352 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Test.Common
+{
+    using System;
+    using System.Collections.Generic;
+    using System.ComponentModel;
+    using System.Globalization;
+    using System.IO;
+    using System.Linq;
+    using System.Net;
+    using System.Security.Cryptography.X509Certificates;
+    using System.Text;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Devices.Client;
+    using Microsoft.Azure.Devices.Edge.Test.Common.Certs;
+    using Microsoft.Azure.Devices.Edge.Test.Common.Config;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Serilog;
+
+    public class LeafDevice
+    {
+        readonly DeviceClient client;
+        readonly Device device;
+        readonly IotHub iotHub;
+        readonly string messageId;
+
+        LeafDevice(Device device, DeviceClient client, IotHub iotHub)
+        {
+            this.client = client;
+            this.device = device;
+            this.iotHub = iotHub;
+            this.messageId = Guid.NewGuid().ToString();
+        }
+
+        public static Task<LeafDevice> CreateAsync(
+            string leafDeviceId,
+            Protocol protocol,
+            AuthenticationType auth,
+            Option<string> parentId,
+            bool useSecondaryCertificate,
+            EdgeCertificateAuthority edgeCa,
+            IotHub iotHub,
+            CancellationToken token)
+        {
+            return Profiler.Run(
+                async () =>
+                {
+                    ITransportSettings transport = protocol.ToTransportSettings();
+                    Platform.InstallEdgeCertificates(edgeCa.Certificates.TrustedCertificates, transport);
+
+                    string edgeHostname = Dns.GetHostName().ToLower();
+
+                    switch (auth)
+                    {
+                        case AuthenticationType.Sas:
+                            return await CreateWithSasAsync(
+                                leafDeviceId,
+                                parentId,
+                                iotHub,
+                                transport,
+                                edgeHostname,
+                                token);
+
+                        case AuthenticationType.CertificateAuthority:
+                            {
+                                string p = parentId.Expect(() => new ArgumentException());
```

Comment:
For learning, why parent Id is required for Cert-related authentication types (CertificateAuthority and SelfSigned)?  I remember at least self-signed cert, I don't need to pass parent id when create leaf device, is it right?

Reply:
Establishing a parent-child relationship (aka scope) for edge/leaf devices is not strictly required when the leaf devices are using SAS, but it's recommended, especially to enable offline support. And when they're using certs (CA or self-signed), it's required. Edge Hub will connect to IoT Hub on behalf of a leaf device using SAS, so if the leaf device didn't supply the key then the parent-child relationship is necessary so the edge can download the key from the cloud.

Reply:
does it apply to modules as well?

Reply:
I don't know for certain, but I don't think so. Modules are already scoped to a device by definition (by means other than the `scope` property of the device), and by default the edge device is responsible for adding/removing module identities when it receives a deployment.

Reply:
Thanks.